nengo/linear_system.py

```diff
@@ -0,0 +1,369 @@
+        # x2 = A2.x2 + B2.u2 = A2.x2 + B2.C1.x1 + B2.D1.u1
+        # y2 = C2.x2 + D2.u2 = C2.x2 + D2.C1.x1 + D2.D1.u1
+        A[:s1, :s1] = other.A
+        A[s1:, :s1] = self.B.dot(other.C)
+        A[s1:, s1:] = self.A
+        B[:s1, :] = other.B
+        B[s1:, :] = self.B.dot(other.D)
+        C[:, :s1] = self.D.dot(other.C)
+        C[:, s1:] = self.C
+        D[:, :] = self.D.dot(other.D)
+
+        x0 = 0
+        if state_size > 0:
+            x0shape = (
+                ()
+                if self.x0.ndim <= 1 and other.x0.ndim <= 1
+                else self.x0.shape[:-1]
+                if self.x0.ndim > other.x0.ndim
+                else other.x0.shape[:-1]
+            )
+            x0a = self._expand_x0(shape=x0shape)
+            x0b = other._expand_x0(shape=x0shape)
+            x0 = np.concatenate([x0b, x0a], axis=-1)
+
+        return cls(
+            sys=(A, B, C, D),
+            analog=self.analog,
+            method=self.method,
+            x0=x0,
+            default_dt=self.default_dt,
+        )
+
+    def discrete_ss(self, dt):
+        """Returns the discrete system state-space representation.
+
+        Note that if the system is ``not analog``, the stored (A, B, C, D) is returned.
+        This function cannot change the ``dt`` of an already discrete system.
+        """
+        A, B, C, D = self.ss
+
+        # discretize (if len(A) == 0, filter is stateless and already discrete)
+        if self.analog and len(A) > 0:
+            A, B, C, D, _ = cont2discrete((A, B, C, D), dt, method=self.method)
+
+        return A, B, C, D
+
+    def discretize(self, dt):
+        """Returns an equivalent discrete linear system.
+
+        Note that if the system is ``not analog``, a copy of the system is returned.
+        This function cannot change the ``dt`` of an already discrete system.
+        """
+        return self._CombineClass(
+            self.discrete_ss(dt),
+            analog=False,
+            method=self.method,
+            x0=self.x0,
+            default_dt=self.default_dt,
+        )
+
+    def make_state(self, shape_in, shape_out, dt, dtype=None, x0=None):
+        dtype = rc.float_dtype if dtype is None else np.dtype(dtype)
+        if dtype.kind != "f":
+            raise ValidationError(
+                "Only float data types are supported (got %s). Please cast "
+                "your data to a float type." % dtype,
+                attr="dtype",
+                obj=self,
+            )
+
+        # create state memory variable X
+        assert shape_in[-1] == self.input_size
+        assert shape_out[-1] == self.output_size
+        assert shape_in[:-1] == shape_out[:-1]
+        X = np.zeros(shape_out[:-1] + (self.state_size,), dtype=dtype)
+        X[:] = self._expand_x0(x0, shape=shape_out[:-1], broadcast=True)
+        return {"X": X}
+
+    def make_step(self, shape_in, shape_out, dt, rng, state):  # noqa: C901
```

Comment:
Is there a way to have only one `make_step` for the linear system? It seems right now there's two (this one, and the old one in `synapses.py`). Can they be combined, e.g., by making the synapse call out to this one? Or not for some reason?

Reply:
This one is fully general, in that it supports systems with any input/output size. The step functions on `LinearFilter` have been optimized for filtering (where the input and output size are both equal to 1). There's also some even more specific functions for cases like only one state, and no state. `LinearFilter` could use this `make_step`, but I think it would be much less efficient (particularly for synapses with one state like `Lowpass`).